plugins/inputs/p4runtime/p4runtime.go

```diff
@@ -0,0 +1,240 @@
+package p4runtime
+
+import (
+	"context"
+	"crypto/tls"
+	_ "embed"
+	"fmt"
+	"io"
+	"strconv"
+	"sync"
+
+	p4ConfigV1 "github.com/p4lang/p4runtime/go/p4/config/v1"
+	p4v1 "github.com/p4lang/p4runtime/go/p4/v1"
+	"golang.org/x/exp/slices"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/influxdata/telegraf"
+	internaltls "github.com/influxdata/telegraf/plugins/common/tls"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+//go:embed sample.conf
+var sampleConfig string
+
+const (
+	defaultDeviceID = "1"
+	defaultAddress  = "127.0.0.1:9559"
+)
+
+type P4runtime struct {
+	Address      string          `toml:"endpoint"`
+	DeviceID     string          `toml:"device_id"`
+	CounterNames []string        `toml:"counter_names"`
+	Log          telegraf.Logger `toml:"-"`
+	EnableTLS    bool            `toml:"enable_tls"`
+	internaltls.ClientConfig
+
+	conn           *grpc.ClientConn
+	client         p4v1.P4RuntimeClient
+	deviceIDParsed uint64
+	wg             sync.WaitGroup
+}
+
+func (*P4runtime) SampleConfig() string {
+	return sampleConfig
+}
+
+func (p *P4runtime) Start(telegraf.Accumulator) error {
+	if p.DeviceID == "" {
+		p.Log.Debugf("Using default deviceID: %v", defaultDeviceID)
+		p.DeviceID = defaultDeviceID
+	}
+	deviceID, err := strconv.ParseUint(p.DeviceID, 10, 64)
+	if err != nil {
+		return err
+	}
+	p.deviceIDParsed = deviceID
+
+	if p.Address == "" {
+		p.Log.Debugf("Using default Address: %v", defaultAddress)
+		p.Address = defaultAddress
+	}
+
+	return p.newP4RuntimeClient()
+}
+
+func (p *P4runtime) Gather(acc telegraf.Accumulator) error {
+	p4Info, err := p.getP4Info()
+	if err != nil {
+		return err
+	}
+
+	if len(p4Info.Counters) == 0 {
+		p.Log.Warn("No counters available in P4 Program!")
```

Comment:
This warning would show up if a query is done and no counters are returned? Are counters added dynamically at a later date? Or would this mean the current endpoint would never have data?

Reply:
P4 program can be changed dynamically, at the runtime (using P4 runtime server that we connect to), so the current endpoint may have data later